source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -95,93 +101,115 @@ namespace aspect
+                      {
+                        double slope_change_ratio = max_total_slope/current_total_slope;
+                        for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                          c[property_index][i] *= slope_change_ratio;
+                      }
                   }
-                else
-                  {
-                    interpolated_value += c[property_index][3] * relative_support_point_location[2] +
-                                          c[property_index][4] * relative_support_point_location[0] * relative_support_point_location[1] +
-                                          c[property_index][5] * relative_support_point_location[0] * relative_support_point_location[2] +
-                                          c[property_index][6] * relative_support_point_location[1] * relative_support_point_location[2] +
-                                          c[property_index][7] * relative_support_point_location[0] * relative_support_point_location[1] * relative_support_point_location[2];
-                  }
-
-                // Overshoot and undershoot correction of interpolated particle property.
-                if (use_global_min_max_limiter)
+                std::size_t positions_index = 0;
+                for (typename std::vector<Point<dim>>::const_iterator itr = positions.begin(); itr != positions.end(); ++itr, ++positions_index)
                   {
-                    interpolated_value = std::min(interpolated_value, global_maximum_particle_properties[property_index]);
-                    interpolated_value = std::max(interpolated_value, global_minimum_particle_properties[property_index]);
+                    Point<dim> relative_support_point_location = mapping.transform_real_to_unit_cell(found_cell, *itr);
+                    double interpolated_value = c[property_index][0];
+                    for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                      {
+                        relative_support_point_location[i - 1] -= unit_offset;
+                        interpolated_value += c[property_index][i] * relative_support_point_location[i - 1];
+                      }
+                    if (use_linear_least_squares_limiter[property_index] == true)
+                      {
+                        // If the limiter is working correctly, we should not be significantly more than machine error outside of the range of the property bounds
+                        Assert((property_minimums[property_index] - interpolated_value) <= (property_maximums[property_index] - property_minimums[property_index]) * 16 * std::numeric_limits<double>::epsilon(), ExcInternalError());
+                        Assert((interpolated_value - property_maximums[property_index]) <= (property_maximums[property_index] - property_minimums[property_index]) * 16 * std::numeric_limits<double>::epsilon(), ExcInternalError());
```

Comment:
This is hard to read. What was wrong with:
```suggestion
                        Assert(interpolated_value >= property_minimums[property_index] - std::abs(property_minimums[property_index]) * 10. * std::numeric_limits<double>::epsilon(), ExcInternalError());
                        Assert(interpolated_value <= property_maximums[property_index] + std::abs(property_maximums[property_index])) * 10. * std::numeric_limits<double>::epsilon(), ExcInternalError());
```

Reply:
The problem with that is the error in interpolated value is going to be impacted by the largest value, not just the closest property bound. If for some reason we had a cell that had a range of [1, 1e15], we would expect that the error would be closer to epsilon * 1e15 than epsilon * 1. I have now implemented a slight modification on your suggestion though and I believe it is ready for review.